integration_test/integration-test.go

```diff
@@ -0,0 +1,324 @@
+		fmt.Println("Server shutdown failed:", err)
+		fail()
+	}
+	close(suite.requestReceived)
+	close(suite.bodyStringReceived)
+}
+
+/*
+* Test if the proxy and the test server has been properly connected
+ */
+func (suite TestSuite) CheckProxyConnection() {
+	// wait and test that the proxy is up
+	fmt.Println("Check proxy connection")
+	time.Sleep(1 * time.Second)
+	response, err := http.Get(suite.proxyURL)
+	if err != nil {
+		fmt.Println(err)
+		fmt.Println("Client connecting to proxy ...")
+		suite.CheckProxyConnection()
+		return
+	}
+	defer response.Body.Close()
+
+	// check that proxy reply indeed comes from server
+	responseBody, _ := ioutil.ReadAll(response.Body)
+	if response.StatusCode != 200 || string(responseBody) != suite.serverMessage {
+		fmt.Println("Proxy connection failed")
+		fmt.Println("Received response", response.Status, string(responseBody))
+		fail()
+	}
+	<-suite.requestReceived
+	<-suite.bodyStringReceived
+	fmt.Println("Client connected to server through proxy")
+}
+
+/*
+* evaluate if a given request is handled properly by the proxy
+* true if the test passed, false if not
+ */
+func (suite TestSuite) Eval(test TestRequest) bool {
+	// initialize request
+	requestBody := test.body.Encode()
+	requestPathParams := test.pathParams.Encode()
+	request, _ := http.NewRequest(
+		"POST",
+		suite.proxyURL+test.path+"?"+requestPathParams,
+		strings.NewReader(requestBody),
+	)
+	request.Header.Set("content-type", "application/x-www-form-urlencoded")
+	for key, val := range test.headers {
+		request.Header.Set(key, val)
+	}
+	for key, val := range test.cookies {
+		request.AddCookie(&http.Cookie{Name: key, Value: val})
+	}
+
+	// send request
+	client := &http.Client{}
+	response, err := client.Do(request)
+	if err != nil {
+		fmt.Println("Send request failed")
+		fmt.Println(err)
+		return false
+	}
+	defer response.Body.Close()
+
+	// wait for the received request to send through
+	time.Sleep(100 * time.Millisecond)
+
+	// get request from server
+	var received *http.Request
+	var receivedBody string
+	select {
+	case received = <-suite.requestReceived:
+	default:
+		received = nil
+	}
+	select {
+	case receivedBody = <-suite.bodyStringReceived:
+	default:
+		receivedBody = ""
+	}
+
+	// verify the request and the response
+	if test.sqli {
+		// if the request contains SQL injection, it should be blocked
+		// if the SQL injection is in body, the server should receive request with empty body
+		// if the SQL injection is in header, the server should never receive any request
+		if test.sqli_body {
+			if receivedBody != "" {
+				fmt.Println("Request with SQL injection in body was not blocked")
+				fmt.Println(receivedBody)
+				return false
+			}
+		} else if received != nil {
+			fmt.Println("Request with SQL injection was not blocked")
+			fmt.Println(response.Status)
+			return false
+		}
+
+		// the client should receive a response from the proxy with status 403 Forbidden
+		if response.StatusCode != 403 {
```

Comment:
Thanks for the example! 

For each request there are three cases: 
1. Without SQL injection 
- In this case, the server should receive the request intact, and the client should receive a http 200 response.
2. Contains SQL injection in header
- In this case, the server should not receive anything, and the client should receive a http 403 response.
3. Contains SQL injection in body, but not in header
- In this case, the server should receive a request with empty body (since the header cannot be buffered, it will be forwarded to the server before the body is checked), and the client should receive a http 403 response.

Should I explicitly express the expected interaction between the server and the client at the top level as in Config of main.go, to the detail of the exact status code? Can I just use an enum to distinguish the three cases, and explain the expected behavior of the cases in comments? In the authorization test, I saw that there's also only a boolean indicating whether the path should be allowed or denied.